lib/babe/babe.go

```diff
@@ -68,7 +68,101 @@ type ServiceConfig struct {
 	Telemetry          telemetry.Client
 }
 
-// NewService returns a new Babe Service using the provided VRF keys and runtime
+// Validate returns error if config does not contain required attributes
+func (sc *ServiceConfig) Validate() error {
+	if sc.Keypair == nil && sc.Authority {
+		return errNoBABEAuthorityKeyProvided
+	}
+
+	if sc.BlockState == nil {
+		return ErrNilBlockState
+	}
+
+	if sc.EpochState == nil {
+		return errNilEpochState
+	}
+
+	if sc.BlockImportHandler == nil {
+		return errNilBlockImportHandler
+	}
+
+	return nil
+}
+
+// ServiceIFace interface that defines methods available to BabeService
+type ServiceIFace interface {
+	Start() error
+	Stop() error
+	EpochLength() uint64
+	Pause() error
+	IsPaused() bool
+	Resume() error
+	SlotDuration() uint64
+}
+
+// ServiceBuilder interface to define the building of babe service
+type ServiceBuilder interface {
+	NewServiceIFace(cfg *ServiceConfig) (ServiceIFace, error)
+}
+
+var _ ServiceBuilder = (*Builder)(nil)
+
+// Builder struct to hold babe builder functions
+type Builder struct{}
+
+//NewServiceIFace returns a new Babe Service using the provided VRF keys and runtime
+func (Builder) NewServiceIFace(cfg *ServiceConfig) (ServiceIFace, error) {
```

Comment:
NewServiceIFace is also an interface so that we can mock it's functions for testing, we're using mocks of this in  dot/services_test.go Test_nodeBuilder_createBABEService.  

I'm not sure using SerivceIFace in rpcServiceSettings is an issue since babe.SerivceIFace implements the functions needed by modules.BlockProducerAPI.

Reply:
I think `Service` should be an interface to allow us to test easily.  I attempted that for `state.Service` but it meant touching a lot of code #2315 .  I think this approach is fine, with eyes on our future goal of overhauling the service interface entirely.